### PR TITLE
Initial Bluetooth LE Central/Client implementation

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description

Initial implementation of Bluetooth Central/Client.

This covers the base implementation for : - 
- Watching for Bluetooth Server devices
- Creating a BluetoothLEDevice object for the device
- Connecting & Discovering available services from device
- Discovering available characteristics in a service
- Discovering available descriptors in a characteristic
- Read / Write to an attribute ( Characteristic/Descriptor)
- Notify events for changes in attributes

See nanoFramework.Device.Bluetooth for full details/examples code

Currently no support for pairing/bonding and authentication. This will be in next PR

There is still a fair amount of commented/disabled code and you may also find some TODO comments

## Motivation and Context
- Resolves nanoFramework/Home#981

## How Has This Been Tested?

Tested with sample code in separate PR.  Will be available once main parts merges and nuget available.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [x] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEDevice.cpp [hunk 1]
```diff
@@ -0,0 +1,1009 @@
...
+    {
+        // Handled
+        return true;
+    }
+
+    // Not handled
+    return false;
+}
+
+int central_gap_event(struct ble_gap_event *event, void *arg)
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEDevice.cpp:157]: Name starting with upper case please.
On the function naming too: we keep having a mix of camel case and snake case naming on the functions. Can we please stick with the camel case?

**Review by josesimoes — COMMENTED** [review]
WOW!! Lot's of work here. Great job. 👏🏻 
Just some minor changes requested. 😅

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothNanoDevice.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
+//-----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothNanoDevice.cpp:1]: fixed

**Review by alberk8 — COMMENTED** [review]
https://github.com/nanoframework/nf-interpreter/blob/189eaa88e646b94c417a4e4810ad58bfa408c145/targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/nimble_utils.cpp#L85

I need to change this so that custom uuid to Guid would work.
`int map[16]{12, 13, 14, 15, 10, 11, 8, 9, 7, 6, 5, 4, 3, 2, 1, 0};`

**Comment by AdrianSoundy** [comment]
> https://github.com/nanoframework/nf-interpreter/blob/189eaa88e646b94c417a4e4810ad58bfa408c145/targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/nimble_utils.cpp#L85
> 
> I need to change this so that custom uuid to Guid would work. `int map[16]{12, 13, 14, 15, 10, 11, 8, 9, 7, 6, 5, 4, 3, 2, 1, 0};`

Can you please explain where you are seeing a problem. I tried BluetoothLEDevcie.GetGattServices() and it returns correct UUID.  Also the line you are changing is not called by anything so it won't make any difference.  I made the GuidToNimbleUUID() method but didn't change the code to call it.
@alberk8

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEAdvertisementWatcher.cpp [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+#define BluetoothLEAdvertisement                                                                                       \
+    Library_sys_dev_ble_native_nanoFramework_Device_Bluetooth_Advertisement_BluetoothLEAdvertisement
+
+bool bleScanActive = false;
+
+//
+// Initiates the GAP general discovery procedure.
+//
+void bleCentralStartScan()
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEAdvertisementWatcher.cpp]: Changed all to Camel case

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEDevice.cpp (hunk 1 above)
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEDevice.cpp]: Fixed

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEDevice.cpp [hunk 4]
```diff
@@ -0,0 +1,1009 @@
...
+#define connectTimeout 7000
+#define asyncTimeout   20000
+
+#define GattDeviceService                                                                                              \
+    Library_sys_dev_ble_native_nanoFramework_Device_Bluetooth_GenericAttributeProfile_GattDeviceService
+#define GattCharacteristic                                                                                             \
+    Library_sys_dev_ble_native_nanoFramework_Device_Bluetooth_GenericAttributeProfile_GattCharacteristic
+#define GattDescriptor Library_sys_dev_ble_native_nanoFramework_Device_Bluetooth_GenericAttributeProfile_GattDescriptor
+
+enum centralOp
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_BluetoothLEDevice.cpp:17]: So enum should be **e_CentralOp**
@josesimoes

**Comment by alberk8** [comment]
> > https://github.com/nanoframework/nf-interpreter/blob/189eaa88e646b94c417a4e4810ad58bfa408c145/targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/nimble_utils.cpp#L85
> > 
> > I need to change this so that custom uuid to Guid would work. `int map[16]{12, 13, 14, 15, 10, 11, 8, 9, 7, 6, 5, 4, 3, 2, 1, 0};`
> 
> Can you please explain where you are seeing a problem. I tried BluetoothLEDevcie.GetGattServices() and it returns correct UUID. Also the line you are changing is not called by anything so it won't make any difference. I made the GuidToNimbleUUID() method but didn't change the code to call it. @alberk8

Sorry for pointing to the wrong line, it should be line 25 under the function NimbleUUID128ToGuid.  Below is what I get from running the Ble Central and displaying the UUID of the services and Characteristic of the [BluetoothLESample3](https://github.com/nanoframework/Samples/tree/main/samples/Bluetooth/BluetoothLESample3).  The incorrect one is only the Custom Services and Characteristics.   The correct uuid should be

```
Guid serviceUuid = new Guid("A7EEDF2C-DA87-4CB5-A9C5-5151C78B0057");
Guid readStaticCharUuid = new Guid("A7EEDF2C-DA89-4CB5-A9C5-5151C78B0057");
```

image

**Comment by AdrianSoundy** [comment]
Looks like a good time to merge now. More things to be done but these can be done in next PR

**Comment by bbhxwl** [comment]
If I scan, stop, connect the client, and continue scanning while connecting, will it affect the connected client? Because I want to connect five different Bluetooth client devices, but I'm not sure when they start up, so I need to scan regularly.

**Comment by josesimoes** [comment]
@bbhxwl this is a question, nothing related with PR workflow. Can you please use our Discord server for these?

**Pull request merged**